%\VignetteIndexEntry{\emph{biosigner} package for molecular signature discovery}
%\VignetteKeywords{biosigner} \\
%\VignettePackage{biosigner} \\

\documentclass[a4paper,11pt]{article}
\usepackage[table]{xcolor}
\usepackage[nottoc, notlof, notlot]{tocbibind}
\setlength{\parindent}{0cm}

<<style-Sweave, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex()
@

\begin{document}
\title{\emph{biosigner}: A new method for signature discovery from omics data}
\author{Philippe Rinaudo and Etienne A. Th\'evenot}
\maketitle

\vspace{1\baselineskip}
\vspace{1\baselineskip}
\vspace{1\baselineskip}
\vspace{1\baselineskip}
\vspace{1\baselineskip}


\tableofcontents

\pagebreak

\setkeys{Gin}{width=0.6\textwidth} 


\section{Introduction}

High-throughput, non-targeted, technologies such as transcriptomics, proteomics and metabolomics, are widely used to discover molecules which allow to efficiently discriminate between biological or clinical conditions of interest (e.g., disease vs control states). Powerful machine learning approaches such as Partial Least Square Discriminant Analysis (PLS-DA), Random Forest (RF) and Support Vector Machines (SVM) have been shown to achieve high levels of prediction accuracy. Feature selection, i.e., the selection of the few features (i.e., the molecular signature) which are of highest discriminating value, is a critical step in building a robust and relevant classifier (\cite{Guyon2003}): First, dimension reduction is usefull to limit the risk of overfitting and reduce the prediction variability of the model; second, intrepretation of the molecular signature is facilitated; third, in case of the development of diagnostic product, a restricted list is required for the subsequent validation steps (\cite{Rifai2006})

Since the comprehensive analysis of all combinations of features is not computationally tractable, several selection techniques have been described, including \emph{filter} (e.g., \textit{p}-values thresholding), \emph{wrapper} (e.g., recursive feature elimination), and \emph{embedded} (e.g., sparse PLS) approaches (\cite{Saeys2007}). The major challenge for such methods is to be fast and extract restricted and stable molecular signatures which still provide high performance of the classifier (\cite{Gromski2014}; \cite{Determan2015}).

\section{The biosigner package}

The \emph{biosigner} implements a new \emph{wrapper} feature selection algorithm: 
\begin{enumerate}
\item the dataset is split into training and testing subsets (by bootstraping, controling class proportion),
\item model is trained on the training set and balanced accuracy is evaluated on the test set,
\item the features are ranked according to their importance in the model,
\item the relevant feature subset at level \emph{f} is found by a binary search: a feature subset is considered relevant if and only if, when randomly permuting the intensities of other features in the test subsets, the proportion of increased or equal prediction accuracies is lower than a defined threshold \emph{f},
\item the dataset is restricted to the selected features and steps 1 to 4 are repeated until the selected list of features is stable.
\end{enumerate}

Three binary classifiers have been included in \emph{biosigner}, namely PLS-DA, RF and SVM, as the performances of each machine learning approach may vary depending on the structure of the dataset (\cite{Determan2015}). The algorithm  returns the \emph{tier} of each feature for the selected classifer(s): tier \emph{S} corresponds to the final signature, i.e., features which have been found significant in all the selection steps; features with tier \emph{A} have been found significant in all but the last selection, and so on for tier \emph{B} to \emph{D}. Tier \emph{E} regroup all previous round of selection.

As for a classical classification algorithm, the \texttt{biosign} method takes as input the \texttt{x} samples times features data frame (or matrix) of intensities, and the \texttt{y} factor (or character vector) of class labels (note that only binary classification is currently available). It returns the signature (\texttt{signatureLs}: selected feature names) and the trained model (\texttt{modelLs}) for each of the selected classifier. The \texttt{plot} method for \texttt{biosign} objects enable to visualize the individual boxplots of the selected features. Finally, the \texttt{predict} method allows to apply the trained classifier(s) on new datasets.
 
The algorithm has been successfully applied to transcriptomics and metabolomics data (\cite{Rinaudosubmitted}; see also the \emph{Hands-on} section below).

\section{Hands-on}

\subsection{Loading}

We first load the \emph{biosigner} package:

<<>>=
library(biosigner)
@

We then use the \emph{diaplasma} metabolomics dataset (\cite{Rinaudosubmitted}) which results from the analysis of plasma samples from 69 diabetic patients were analyzed by reversed-phase liquid chromatography coupled to high-resolution mass spectrometry (Orbitrap Exactive) in the negative ionization mode. The raw data were pre-processed with XCMS and CAMERA (5,501 features), corrected for signal drift, log10 transformed, and annotated with an in-house spectral database. The patient's age, body mass index, and diabetic type are recorded.

<<>>=
data(diaplasma)
@

We attach \emph{diaplasma} to the search path and display a summary of the content of the \emph{dataMatrix}, \emph{sampleMetadata} and \emph{variableMetadata} with the \emph{strF} function (from the \emph{ropls} package):

<<>>=
attach(diaplasma)
strF(dataMatrix)
strF(sampleMetadata)
strF(variableMetadata)
@

We see that the \emph{diaplasma} list contains three objects:
\begin{description}
\item[dataMatrix] 69 samples x 5,501 matrix of numeric type containing the intensity profiles (log10 transformed),
\item[sampleMetadata] a 69 x 3 data frame, with the patients'
\begin{description}
\item[type] diabetic type, factor
\item[age] numeric
\item[bmi] body mass index, numeric
\end{description}
\item[variableMetadata] a 5,501 x 8 data frame, with the median m/z ('mzmed', numeric) and the median retention time in seconds ('rtmed', numeric) from XCMS, the 'isotopes' (character), 'adduct' (character) and 'pcgroups' (numeric) annotations from CAMERA, the names of the m/z and RT matching compounds from an in-house spectra of commercial metabolites ('name\textunderscore hmdb', character), and the \textit{p}-values resulting from the non-parametric hypothesis testing of difference in medians between types ('type\textunderscore wilcox\textunderscore fdr', numeric), and correlation with age ('age\textunderscore spearman\textunderscore fdr', numeric) and body mass index ('bmi\textunderscore spearman\textunderscore fdr', numeric), all corrected for multiple testing (False Discovery Rate).
\end{description}

We can observe that the 3 clinical covariates (diabetic \emph{type}, \emph{age}, and \emph{bmi}) are stronlgy associated (Figure \ref{Fig:covariate}):

<<eval=FALSE>>=
with(sampleMetadata, plot(age, bmi, cex = 1.5, col = ifelse(type == "T1", "blue", "red")))
legend("topleft", cex = 1.5, legend = paste0("T", 1:2), text.col = c("blue", "red"))
@

<<echo=FALSE,results=hide>>=
pdf("biosigner-figCovariate.pdf")
with(sampleMetadata, plot(age, bmi, cex = 1.5, col = ifelse(type == "T1", "blue", "red")))
legend("topleft", cex = 1.5, legend = paste0("T", 1:2), text.col = c("blue", "red"))
dev.off()
@

\begin{figure}[h!]
	\begin{center}
		\includegraphics[width=4in]{biosigner-figCovariate.pdf}
	\end{center}
	\caption{\emph{age}, \emph{body mass index (bmi)} and diabetic \emph{type} (T1 and T2) of the patients from the \emph{diaplasma} cohort.}
	\label{Fig:covariate}
\end{figure}

\subsection{Molecular signatures}

Let us look for signatures of \emph{type} in the \emph{diaplasma} dataset by using the \texttt{biosign} method. To speed up computations in this demo vignette, we restrict the number of features (from 5,501 to about 500) and the number of bootstraps (5 instead of 50 [default]); the selection on the whole dataset, 50 bootstraps, and the 3 classifiers, takes around 10 min.

<<>>=
featureSelVl <- variableMetadata[, "mzmed"] >= 450 & variableMetadata[, "mzmed"] < 500
sum(featureSelVl)
dataMatrix <- dataMatrix[, featureSelVl]
variableMetadata <- variableMetadata[featureSelVl, ]
@

<<eval=FALSE>>=
set.seed(123)
diaSign <- biosign(dataMatrix, sampleMetadata[, "type"], bootI = 5)
set.seed(NULL)
@

<<echo=FALSE>>=
set.seed(123)
diaSign <- biosign(dataMatrix, sampleMetadata[, "type"], bootI = 5, plotL = FALSE)
set.seed(NULL)
@

<<echo=FALSE>>=
plot(diaSign, file.pdfC = "biosigner-figTier.pdf")
@

\begin{figure}[h!]
	\begin{center}
		\includegraphics[width=4in]{biosigner-figTier.pdf}
	\end{center}
	\caption{Relevant signatures for the PLS-DA, Random Forest, and SVM classifiers extracted from the \emph{diaplasma} dataset. The (\emph{S}) tier corresponds to the final metabolite signature, i.e., metabolites which passed through all the selection steps.}
	\label{Fig:Tier}
\end{figure}

The arguments are:

\begin{description}
\item[x] the numerical matrix (or data frame) of intensities (samples as rows, variables as columns),
\item[y] the factor (or character) specifying the sample labels from the 2 classes,
\item[methodVc] the classifier(s) to be used; here, the default \texttt{all} value means that all classifiers available (\texttt{plsda}, \texttt{randomforest}, and \texttt{svm}) are selected,
\item[bootI] the number of bootstraps is set to 5 to speed up computations when generating this vignette; we however recommend to keep the default 50 value for your analyzes (otherwise signatures may be less stable).
\end{description}

Note:
\begin{itemize}
\item If some features from the \texttt{x} matrix/data frame contain missing values (NA), these features will be removed prior to modeling with Random Forest and SVM (in contrast, the NIPALS algorithm from PLS-DA can handle missing values),
\item The \texttt{set.seed} command was used here to be sure that the results from this vignette can be reproduced exactly; by choosing alternative seeds (and the default \texttt{bootI = 50}), similar signatures are obtained, showing the stability of the selection.
\end{itemize}

The resulting signatures for the 3 selected classifiers are both printed and plotted (Figure \ref{Fig:Tier}) as \emph{tiers} from \emph{S}, \emph{A}, up to \emph{E} by decreasing relevance. The (\emph{S}) tier corresponds to the final signature, i.e. features which passed through all the backward selection steps. In contrast, features from the other tiers were discarded during the last (\emph{A}) or previous (\emph{B} to \emph{E}) selection rounds.

Note that \texttt{tierMaxC = 'A'} argument in the \texttt{print} and \texttt{plot} methods can be used to view the features from the larger \emph{S+A} signatures (especially when no \emph{S} features have been found, or when the performance of the \emph{S} model is much lower than the \emph{S+A} model).

The performance of the model built with the input dataset (\emph{balanced accuracy}: mean of the \emph{sensitivity} and \emph{specificity}), or the subset restricted to the \emph{S} or \emph{S+A} signatures are shown. We see that with 1 to 5 \emph{S} feature signatures (i.e., less than 1\% of the input), the 3 classifiers achieve good performances (even higher than the full Random Forest and SVM models). Furthermore, reducing the number of features decreases the risk of building non-significant models (i.e., models which do not perform significantly better than those built after randomly permuting the labels). The signatures from the 3 classifiers have some distinct features, which highlights the interest of comparing various machine learning approaches.

The individual boxplots of the features from the \emph{complete} signature can be visualized with (Figure \ref{Fig:Boxplot}):

<<eval=FALSE>>=
plot(diaSign, typeC = "boxplot")
@

<<echo=FALSE>>=
plot(diaSign, typeC = "boxplot", file.pdfC = "biosigner-figBoxplot.pdf")
@

\begin{figure}[h!]
	\begin{center}
		\includegraphics[width=4in]{biosigner-figBoxplot.pdf}
	\end{center}
	\caption{Individual boxplots of the features selected for at least one of the classification methods. Features selected for a single classifier are colored (red for PLS-DA, green for Random Forest and blue for SVM).}
	\label{Fig:Boxplot}
\end{figure}

Let us see the metadata of the \emph{complete} signature:

<<>>=
variableMetadata[getSignatureLs(diaSign)[["complete"]], ]
@

We observe that the taurochenodeoxycholic acid has been annotated, in addition to another [M-H]- ion at 470.233 Da. Six out of the 8 features are very significant by univariate hypothesis testings of difference between \emph{type} medians, and to a lesser extent, of the correlation with \emph{age} and \emph{body mass index}.

\subsection{Predictions}

Let us keep split the dataset into a training (the first 4/5th of the 183 samples) and a testing subsets, and extract the relevant features from the training subset:

<<>>=
trainVi <- 1:floor(0.8 * nrow(dataMatrix))
testVi <- setdiff(1:nrow(dataMatrix), trainVi)
@

<<eval=FALSE>>=
set.seed(123)
diaTrain <- biosign(dataMatrix[trainVi, ], sampleMetadata[trainVi, "type"], bootI = 5)
set.seed(NULL)
@                    

<<echo=FALSE>>=
set.seed(123)
diaTrain <- biosign(dataMatrix[trainVi, ], sampleMetadata[trainVi, "type"], bootI = 5, plotL = FALSE)
set.seed(NULL)
@

We extract the fitted types on the training dataset restricted to the \emph{S} signatures:

<<>>=
diaFitDF <- predict(diaTrain)
@

We then print the confusion tables for each classifier:

<<>>=
lapply(diaFitDF, function(predFc) table(actual = sampleMetadata[trainVi,
"type"], predicted = predFc))
@

and the corresponding balanced accuracies:

<<>>=
sapply(diaFitDF, function(predFc) { conf <- table(sampleMetadata[trainVi,
"type"], predFc)
conf <- sweep(conf, 1, rowSums(conf), "/")
round(mean(diag(conf)), 3)
})
@

Note that these values are slightly different from the accuracies returned by \texttt{biosign} because the latter are computed by using the resampling scheme selected by the \texttt{bootI} (or \texttt{crossvalI}) arguments:

<<>>=
round(getAccuracyMN(diaTrain)["S", ], 3)
@

Finally, we can compute the performances on the test subset:

<<>>=
diaTestDF <- predict(diaTrain, newdata = dataMatrix[testVi, ])
sapply(diaTestDF, function(predFc) { conf <- table(sampleMetadata[testVi,
"type"], predFc)
conf <- sweep(conf, 1, rowSums(conf), "/")
round(mean(diag(conf)), 3)
})
@

\subsection{Closing}

Before closing this example session, we detach \emph{diaplasma} from the search path:

<<>>=
detach(diaplasma)
@

\subsection{Session information}

<<eval=FALSE,results=hide>>=
sessionInfo()
@
<<echo=FALSE,results=tex>>=
toLatex(sessionInfo())
@

\section{Extraction of biomarker signatures from other omics datasets}

\subsection{Physiological variations of the human urine metabolome (metabolomics)}

The \emph{sacurine} LC-HRMS dataset from the dependent \emph{ropls} package can also be used (\cite{Thevenot2015}): Urine samples from a cohort of 183 adults were analyzed by using an LTQ Orbitrap in the negative ionization mode. A total of 109 metabolites were identified or annotated at the MSI level 1 or 2. Signal drift and batch effect were corrected, and each urine profile was normalized to the osmolality of the sample. Finally, the data were log10 transformed (see the \emph{ropls} vignette for further details and examples).

We can for instance look for signatures of the \emph{gender}:

<<>>=
data(sacurine)
set.seed(123) ##
sacSign <- biosign(sacurine[["dataMatrix"]], sacurine[["sampleMetadata"]][, "gender"],
methodVc = "plsda")
set.seed(NULL)
@

\subsection{Apples spikes with known compounds (metabolomics)}

The \emph{spikedApples} dataset was obtained by LC-HRMS analysis (SYNAPT Q-TOF, Waters) of one control and three spiked groups of 10 apples each. The spiked mixtures consists in 2 compounds which were not naturally present in the matrix and 7 compounds aimed at achieving a final increase of 20\%, 40\% or 100\% of the endogeneous concentrations. The authors identified 22 features (out of the 1,632 detected in the positive ionization mode; i.e. 1.3\%) which came from the spiked compounds. The dataset is included in the \emph{BioMark} R Bioconductor package (\cite{Franceschi2012}). Let us use the \emph{control} and \emph{group1} samples (20 in total) in this study.

<<>>=
library(BioMark)
data(SpikePos)
group1Vi <- which(SpikePos[["classes"]] %in% c("control", "group1"))
appleMN <- SpikePos[["data"]][group1Vi, ]
spikeFc <- factor(SpikePos[["classes"]][group1Vi])
annotDF <- SpikePos[["annotation"]]
rownames(annotDF) <- colnames(appleMN)
@

We can check that no clear separation can be observed by PCA:

<<>>=
pcaLs <- opls(appleMN, plotL = FALSE)
plot(pcaLs, parAsColFcVn = spikeFc)
@

and that PLS-DA modeling with the full dataset is not significant (as seen on the top left plot: 7 out of 20 models trained after random permutations of the labels have Q2 values higher than the model trained with the true labels):

<<>>=
plsLs <- opls(appleMN, spikeFc)
@

Let us now extract the molecular signatures:

<<eval=FALSE>>=
set.seed(123)
appleSign <- biosign(appleMN, spikeFc)
set.seed(NULL)
@

The \emph{449.1/327} corresponds to the Cyanidin-3-galactoside (absent in the control) and the \emph{475.1/434.7} is probably a potassium adduct of the Phloridzin (80\% concentration increase in group1; \cite{Franceschi2012}).

<<eval=FALSE>>=
annotDF <- SpikePos[["annotation"]]
rownames(annotDF) <- colnames(appleMN)
annotDF[getSignatureLs(appleSign)[["complete"]], c("adduct", "found.in.standards")]
@

\subsection{Bone marrow from acute leukemia patients (transcriptomics)}

Samples from 47 patients with acute lymphoblastic leukemia (ALL) and 25 patients with acute myeloid leukemia (AML) have been analyzed using Affymetrix Hgu6800 chips resulting in expression data of 7,129 gene probes (\cite{Golub1999}). The \emph{golub} dataset is available in the \emph{golubEsets} package from Bioconductor. Let us compute for example the SVM signature (to speed up this demo example, the number of features is restricted to 500):

<<>>=
library(golubEsets)
data(Golub_Merge)
golubMN <- t(exprs(Golub_Merge))
leukemiaFc <- pData(Golub_Merge)[["ALL.AML"]]
table(leukemiaFc)
varSubVi <- 1501:2000
set.seed(123)
golubSign <- biosign(golubMN[, varSubVi], leukemiaFc, methodVc = "svm")
set.seed(NULL)
@

The computation results in a signature of 2 features only and a sparse SVM model performing almost as well (94.4\% accuracy) as the model trained on the dataset of 500 variables (95.0\% accuracy).

The \emph{hu6800.db} bioconductor package can be used to get the annotation of the selected probes (\cite{Carlson}):

<<>>=
library(hu6800.db)
sapply(getSignatureLs(golubSign)[["complete"]],
       function(probeC)
       get(probeC, env = hu6800GENENAME))
@

Cystatin C is part of the 50 gene signature selected by Golub and colleagues on the basis of a metric derived from the Student's statistic of mean differences between the AML and ALL groups (\cite{Golub1999}). Interestingly, the second probe, myeloperoxidase, is a cytochemical marker for the diagnosis (and also potentially the prognosis) of acute myeloid leukemia (AML).

<<eval=FALSE,echo=FALSE>>=
rm(list = ls())
@

\bibliography{biosigner}

\end{document}
